massa-versioning-worker/src/versioning_middleware.rs

```diff
@@ -0,0 +1,156 @@
+use std::collections::{HashMap, VecDeque};
+
+use massa_models::amount::Amount;
+use massa_time::{MassaTime, TimeError};
+
+use crate::versioning::{Advance, MipStore};
+use massa_versioning_exports::VersioningError;
+
+/// Struct used to keep track of announced versions in previous blocks
+pub struct VersioningMiddleware {
+    latest_announcements: VecDeque<u32>,
+    counts: HashMap<u32, usize>,
+    nb_blocks_considered: usize,
+    mip_store: MipStore,
+}
+
+impl VersioningMiddleware {
+    /// Creates a new empty versioning middleware.
+    pub fn new(nb_blocks_considered: usize, mip_store: MipStore) -> Self {
+        VersioningMiddleware {
+            latest_announcements: VecDeque::with_capacity(nb_blocks_considered + 1),
+            counts: HashMap::new(),
+            nb_blocks_considered,
+            mip_store,
+        }
+    }
+
+    /// Add the announced version of a new block
+    /// If needed, remove the info related to a ancient block
+    pub fn new_block(&mut self, version: u32) {
+        self.latest_announcements.push_back(version);
+
+        *self.counts.entry(version).or_default() += 1;
+
+        // If the queue is too large, remove the most ancient block and its associated count
+        if self.latest_announcements.len() > self.nb_blocks_considered {
+            let prev_version = self.latest_announcements.pop_front();
+            if let Some(prev_version) = prev_version {
+                *self.counts.entry(prev_version).or_insert(1) -= 1;
+            }
+        }
+
+        //let _ = self.create_and_send_advance_message(*self.counts.get(&version).unwrap(), version);
+        let _ = self.create_and_send_advance_message_for_all();
+    }
+
+    fn create_and_send_advance_message_for_all(&mut self) -> Result<(), VersioningError> {
+        let mut store = self.mip_store.0.write();
+
+        let now = MassaTime::now().map_err(|_| {
+            VersioningError::ModelsError(massa_models::error::ModelsError::TimeError(
+                TimeError::TimeOverflowError,
+            ))
+        })?;
+
+        // Possible optimisation: filter the store to avoid advancing on failed and active versions
```

Comment:
Let's add a TODO / OPTIM marker here so we do not forget it :)

Reply:
Done